### DISCUSSION TITLE
How to decode the UploadFile when read in async way?

### DISCUSSION THREAD

**Discussion by superxuu** [fastapi/fastapi#9081] [Questions] [answered]
### First check

* [ ] I added a very descriptive title to this issue.
* [ ] I used the GitHub search to find a similar issue and didn't find it.
* [ ] I searched the FastAPI documentation, with the integrated search.
* [ ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI,  UploadFile

@app.post("/uploadfile/")
async def create_upload_file(
file: UploadFile = File(...)
):

    contents = await file.read()
    return {"contents": contents}
```

### Description

my UploadFile  is coded in GBK， bug Fastapi's default is utf8， so when I read the UploadFile  ,I got the error:
***UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd0 in position 0: invalid continuation byte***

how can I decode the UploadFile in my own way?
### Environment

* OS:  Windows
* FastAPI Version : 0.45.0

To know the FastAPI version use:

```bash
python -c "import fastapi; print(fastapi.__version__)"
```

* Python version:Python 3.6.3

To know the Python version use:

```bash
python --version
```

**Comment by sedhha** [comment]
Was it resolved by any chance?

**Comment by liqiujiong** [comment]
i happen same problem in fastapi-0.75.2 that i upload .wav file ，

``` 
# file 参数类型是 UploadFile
@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile = File(...)):
    result = {
        "filename": file.filename,
        "content-type": file.content_type,
        "read": await file.read()
    }
    return result

**Comment by nanarino** [comment]
我在 fastapi0.86.0 中用GB2312的文本文件或者二进制文件测试：
```python
content = await file.read()
print(type(content))  # <class 'bytes'>
```
结果都是bytes， 没有遇到str的情况。我直接用wb+写入后可以正常使用
但是当我开启 vscode 的 type checking on后
会遇到警告：Type "bytes | str" cannot be assigned to type "bytes", "str" is incompatible with "bytes"
🤔

**Reply by nanarino** [reply]
没有问题了。这个在新的版本得到了修复。
现在`file.read`的类型提示是：`(method) def read(size: int = -1) -> Coroutine[Any, Any, bytes]`
😊

**Comment by YuriiMotov** [comment, marked as answer]
Just use `.decode(encoding="gbk")`:

```python
from io import BytesIO
from fastapi import FastAPI, UploadFile
from fastapi.testclient import TestClient

app = FastAPI()

@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile):
    contents = await file.read()
    return {"contents": contents.decode(encoding="gbk")}



def test_upload_file_gbk_encoded():
    # Prepare GBK-encoded content
    text = "你好，世界"  # "Hello, World" in Chinese
    gbk_bytes = text.encode("gbk")

    file_content = BytesIO(gbk_bytes)
    file_content.name = "hello.txt"

    client = TestClient(app)
    response = client.post(
        "/uploadfile/",
        files={"file": ("hello.txt", file_content, "text/plain")},
    )

    assert response.status_code == 200
    assert response.json() == {"contents": text}

```
